library/core/src/array/iter/iter_inner.rs

```diff
@@ -175,6 +175,19 @@ impl<T> PolymorphicIter<[MaybeUninit<T>]> {
         })
     }
 
+    #[inline]
+    pub(super) fn peek_with<U>(&self, func: impl for<'b> FnOnce(Option<&'b T>) -> U) -> U {
+        let tmp = self.alive.clone().next().map(|idx| {
+            // SAFETY: `idx` is in self.alive range
+            unsafe { self.data.get_unchecked(idx).assume_init_read() }
+        });
+
+        let out = func(tmp.as_ref());
+        // Avoid dropping before the item is consumed
+        crate::mem::forget(tmp);
+        out
+    }
```

Comment:
Why do we have to copy the data out of the slice, as opposed to passing a reference (that points directly into the slice) as an argument to the callback?